src/main/java/net/caffeinemc/sodium/render/chunk/occlusion/ChunkOcclusion.java

```diff
@@ -1,136 +1,266 @@
+                // I (northwest x- z-)
+                tryAdd(tree.getSectionId(chunkX - dx, chunkY, chunkZ - dz), tree, inDirection, (byte) (mask | XN | ZN), allowedTraversalDirections, visibleTraversalDirections, visible, inner);
             }
         }
-    
-        int[] sortedIndices = new int[count];
 
-        for (int i = 0; i < count; i++) {
-            sortedIndices[i] = i;
+        for (int distance = 1; distance <= renderDistance; distance++) {
+            // nodes are checked at the following distances:
+            // 1 2 3 . 3 2 1
+            // 2 3 . . . 3 2
+            // 3 . . . . . 3
+            // . . . . . . .
+            // 3 . . . . . 3
+            // 2 3 . . . 3 2
+            // 1 2 3 . 3 2 1
+
+            IntList outer = sections[2 * renderDistance - distance + 1] = new IntArrayList();
+
+            for (int i = 0; i < distance; i++) {
+                int dx = renderDistance - i;
+                int dz = renderDistance - distance + i + 1;
+
+                // J (northeast x+ z-)
+                tryAdd(tree.getSectionId(chunkX + dx, chunkY, chunkZ - dz), tree, inDirection, (byte) (mask | XP | ZN), allowedTraversalDirections, visibleTraversalDirections, visible, outer);
+                // K (southeast x+ z+)
+                tryAdd(tree.getSectionId(chunkX + dx, chunkY, chunkZ + dz), tree, inDirection, (byte) (mask | XP | ZP), allowedTraversalDirections, visibleTraversalDirections, visible, outer);
+                // L (southwest x- z+)
+                tryAdd(tree.getSectionId(chunkX - dx, chunkY, chunkZ + dz), tree, inDirection, (byte) (mask | XN | ZP), allowedTraversalDirections, visibleTraversalDirections, visible, outer);
+                // M (northwest x- z-)
+                tryAdd(tree.getSectionId(chunkX - dx, chunkY, chunkZ - dz), tree, inDirection, (byte) (mask | XN | ZN), allowedTraversalDirections, visibleTraversalDirections, visible, outer);
+            }
         }
 
-        IntArrays.mergeSort(sortedIndices, (k1, k2) -> Float.compare(distances.getFloat(k1), distances.getFloat(k2)));
-    
-        int[] sortedSections = new int[count];
+        // Trim the section lists (so no empty at the front or back)
+        int first = 0;
+        int last = sections.length - 1;
 
-        for (int i = 0; i < count; i++) {
-            sortedSections[i] = sections.getInt(sortedIndices[i]);
+        while (sections[first].isEmpty()) {
+            if (++first >= sections.length) {
+                return null;
+            }
         }
 
-        return IntArrayList.wrap(sortedSections, count);
-    }
+        while (sections[last].isEmpty()) {
+            last--;
+        }
 
-    private static int markDirection(int state, int dir) {
-        return (state | (1 << dir));
-    }
+        last++;
 
-    private static boolean hasDirection(int state, int dir) {
-        return (state & (1 << dir)) != 0;
+        if (first != 0 && last != sections.length) {
```

Comment:
```suggestion
        if (first != 0 && last != sections.length - 1) {
```
At present, this "trim array" path is always run.

Reply:
yes and no, it's a off by one where I initialize.

Reply:
actually no, the `last++` fixes that.

Reply:
Oh, oops, sorry. This `if` just looks fishy.